### DISCUSSION TITLE
Xray-core WireGuard 使用笔记

### DISCUSSION THREAD

**Discussion by chika0801** [XTLS/Xray-core#1915] [General]
存档，原文链接 https://github.com/XTLS/Xray-core/issues/1896

**Comment by testcaoy7** [comment]
Use WireGuard embedded in Xray for whatever purpose is **a very dumb idea**.
Its performance is **poor**. I mean very poor. Don't expect to achieve 1Gbps throughput.

If you want to use a **VPN** then use it correctly. You shall always use WireGuard implemented in kernel.

But you want split tunneling and tunnel only non-China traffic through VPN tunnel.
In that case, OSPF (Open Shortest Path First) is your friend. OSPF is a dynamic routing protocol and since VPN operates at network level, it only make sense that the split tunneling should be done in routing table.

I recommend this project: https://github.com/dndx/nchnroutes
It can auto-generate routing table for non-China ip addresses and export them in `bird` configuration file. 
Install `bird` on your system and all you need to do is to use Xray for **DNS only**. (Yes, Xray can be configured to become a smart DNS server. Hint: use `dokodemo-door` at port 53 and `dns` for outbound.)

**Reply by testcaoy7** [reply]
Use China based DNS server for `geosite:cn` and oversea DNS for the rest.

**Reply by testcaoy7** [reply]
P.P.S.
You can also tunnel WireGuard through Vless and VMess protocol.
Don't use shadowsocks for this, the packet encoding is different and it won't work.

**Comment by chika0801** [comment]
> Use WireGuard embedded in Xray for whatever purpose is a very dumb idea.

In mainland China, if you use WireGuard to connect to VPS outside mainland China, the ports of VPS will be blocked by firewall.

The main purpose of adding WireGuard as an outbound protocol in Xray is to use WireGuard on the VPS (server side) to facilitate access to warp.

One of the purposes of using Warp is to unlock netflix streaming checking, or to allow VPS IPs that would otherwise not support the use of chatgpt to allow the VPS to use chatgpt.

**Reply by testcaoy7** [reply]
I already gave an solution (in `P.P.S.`):
Tunnel your WireGuard connection through Xray (Vless or VMess, with TLS or Reality).
In that case, it will not get blocked.

Cloudflare Warp can be connected via kernel WireGuard as well. You can use `wgcf` project for that. 
I understand in some regions, Warp can only be connected via Xray's WireGuard. That is unfortunate.

**Comment by chika0801** [comment]
> Tunnel your WireGuard connection through Xray (Vless or VMess, with TLS or Reality).
> In that case, it will not get blocked.

According to my personal opinion, in mainland China, the outer layer of the use of such as VLESS + REALITY, inside the use of WireGuard. Not many people use it.

If it is a personal self-built VPS users, either use hysteria tuic and other UDP type of protocol, or VLESS such TCP type of protocol.

> I understand in some regions, Warp can only be connected via Xray's WireGuard. That is unfortunate.

https://github.com/chika0801/Xray-examples/blob/main/wireguard.md

`"reserved":[0, 0, 0]`

According to the experience summarized by netizens in mainland China, the IPs of VPS in Hong Kong, and some VPS in Los Angeles, require the value of the above field in order to connect to Warp.

**Reply by testcaoy7** [reply]
I don't know about Los Angeles but HK, SG, TW do require a "magic number" for Warp service. Japan does not require this and can work with in-kernel WireGuard.

You should try `tunnelling WireGuard via Vless UoT`. I can achieve near 1Gbps throughput  which is far better than built-in WireGuard.

**Reply by ali6043** [reply]
Hi brother.can you send tutorial link for wireguard over vless.thanks

**Comment by chika0801** [comment]
> You should try tunnelling WireGuard via Vless UoT. I can achieve near 1Gbps throughput which is far better than built-in WireGuard.

I personally don't use WireGuard either, I mainly use VLESS+REALITY and hysteria tuic.

**Reply by Saeid008** [reply]
Hi.
Since the Wireguard has been added to Xray-core and now supported, Is that possible Xray support l2tp too? I mean there are many router in market that only support l2tp or pptp. if xray support it, it will be great. I don't know where can I ask this request from Xray developer.
Thanks.

**Reply by chika0801** [reply]
You can make a post in the discussion forum.

As far as I know the developers don't have time to work on new features either.

The wireguard feature is also a non-xray maintainer that submitted a PR that was merged into xray. Please understand some special circumstances of working on this project in this environment in China, even though China has a large population, the number of people focusing on this area is very small.
